Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm

```diff
@@ -337,6 +343,22 @@ JSObjectRef toJSError(JSContextRef context, NSString *callingAPIName, NSString *
     return locale.languageCode;
 }
 
+NSDictionary *toWebAPI(const WebExtensionMessageSenderParameters& parameters)
+{
+    NSMutableDictionary *result = [NSMutableDictionary dictionary];
+
+    if (parameters.tabParameters)
+        result[tabKey] = toWebAPI(parameters.tabParameters.value());
+
+    if (parameters.frameIdentifier)
+        result[frameIdKey] = @(toWebAPI(parameters.frameIdentifier.value()));
+
+    if (parameters.url.isValid())
+        result[urlKey] = (NSURL *)parameters.url;
+
+    return [result copy];
```

Comment:
Also ARC.

Reply:
A `-copy` of an `NSMutableDictionary` is also low-cost because Foundation implements copy-on-write semantics when making a non-mutable copy of a mutable dictionary.